drivers/sensor/nuvoton_adc_cmp_npcx/adc_cmp_npcx.c

```diff
@@ -0,0 +1,272 @@
+	param.val = (uint32_t)&data->work;
+	ret = adc_npcx_threshold_ctrl_set_param(config->adc, config->th_sel,
+						&param);
+	if (ret) {
+		goto init_error;
+	}
+
+	/* Set threshold value if set on device tree */
+	if (config->thr_mv != ADC_CMP_NPCX_UNDEFINED) {
+		param.type = ADC_NPCX_THRESHOLD_PARAM_THVAL;
+		/* Convert from millivolts to ADC raw register value */
+		ret = adc_npcx_threshold_mv_to_thrval(config->thr_mv,
+						&param.val);
+		if (ret) {
+			goto init_error;
+		}
+
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+				config->th_sel,	&param);
+		if (ret) {
+			goto init_error;
+		}
+	}
+
+	/* Set threshold comparison if set on device tree */
+	if (config->comparison == ADC_CMP_NPCX_GREATER ||
+	    config->comparison == ADC_CMP_NPCX_LESS_OR_EQUAL) {
+		param.type = ADC_NPCX_THRESHOLD_PARAM_L_H;
+		param.val =
+			config->comparison == ADC_CMP_NPCX_GREATER ?
+			ADC_NPCX_THRESHOLD_PARAM_L_H_HIGHER :
+			ADC_NPCX_THRESHOLD_PARAM_L_H_LOWER;
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+				config->th_sel,	&param);
+	}
+
+init_error:
+	if (ret) {
+		LOG_ERR("Error setting parameter %d - value %d",
+			(uint32_t)param.type, param.val);
+	}
+
+	return ret;
+}
+
+int adc_cmp_npcx_attr_set(const struct device *dev,
+			  enum sensor_channel chan,
+			  enum sensor_attribute attr,
+			  const struct sensor_value *val)
+{
+	const struct adc_cmp_npcx_config *const config = dev->config;
+	struct adc_npcx_threshold_param param;
+	int ret;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	switch ((uint16_t)attr) {
+	case SENSOR_ATTR_LOWER_THRESH:
+	case SENSOR_ATTR_UPPER_THRESH:
+		/* Set threshold value first */
+		param.type = ADC_NPCX_THRESHOLD_PARAM_THVAL;
+		param.val = val->val1;
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
+		if (ret) {
+			break;
+		}
+
+		/* Then set lower or higher threshold */
+		param.type = ADC_NPCX_THRESHOLD_PARAM_L_H;
+		param.val = attr == SENSOR_ATTR_UPPER_THRESH ?
+			ADC_NPCX_THRESHOLD_PARAM_L_H_HIGHER :
+			ADC_NPCX_THRESHOLD_PARAM_L_H_LOWER;
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
+		break;
+	case SENSOR_ATTR_LOWER_VOLTAGE_THRESH:
+	case SENSOR_ATTR_UPPER_VOLTAGE_THRESH:
+		/* Set threshold value first */
+		param.type = ADC_NPCX_THRESHOLD_PARAM_THVAL;
+		/* Convert from millivolts to ADC raw register value */
+		ret = adc_npcx_threshold_mv_to_thrval(val->val1,
+						&param.val);
+		if (ret) {
+			break;
+		}
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
+		if (ret) {
+			break;
+		}
+
+		/* Then set lower or higher threshold */
+		param.type = ADC_NPCX_THRESHOLD_PARAM_L_H;
+		param.val =
+			(uint16_t)attr == SENSOR_ATTR_UPPER_VOLTAGE_THRESH ?
+			ADC_NPCX_THRESHOLD_PARAM_L_H_HIGHER :
+			ADC_NPCX_THRESHOLD_PARAM_L_H_LOWER;
+
+		ret = adc_npcx_threshold_ctrl_set_param(config->adc,
+						config->th_sel, &param);
```

Comment:
Please add a helper routine for setting the threshold.  The `SENSOR_ATTR_LOWER/UPPER_THRESH` cases are identical to `SENSOR_ATTR_LOWER/UPPER_VOLTAGE_THRESH` except that the voltage threshold cases convert from mv to a register value first.